### PR TITLE
Added GROUP_CONCAT/STRING_AGG support

### PR DESCRIPTION
**Pull request by jimbo2150**
Adds support for the `GROUP_CONCAT()` (known as `STRING_AGG()` in PostgreSQL and MS SQL/tSQL) aggregate functions.

This is a PR for issue #14455

### CODE REVIEW

**Comment by jimbo2150** [comment]
Have an issue where SQLite is returning different results - appears to due to fixtures duplicating data when run more than once (and it does). All the other databases seem to assume DISTINCT on the grouped associations except SQLite. Looking at it deeper I would agree that SQLite is likely in the right whereas the others are assuming deduplication. Since only MySQL supports the DISTINCT modifier within the function expression, would be better to do a subquery across all (with DISTINCT) but am having an issue getting the subquery to be generated rather than a join.

File: src/Database/Expression/AggregateExpression.php [hunk 1]
```diff
@@ -179,9 +184,9 @@ public function sql(ValueBinder $generator): string
         $sql = parent::sql($generator);
         if ($this->window !== null) {
             if ($this->window->isNamedOnly()) {
-                $sql .= ' OVER ' . $this->window->sql($generator);
+                $sql .= " {$this->clause} " . $this->window->sql($generator);
```
Comment (othercorey) [inline src/Database/Expression/AggregateExpression.php:187]: What is the point of this change?

Comment (jimbo2150) [reply src/Database/Expression/AggregateExpression.php:187]: MS SQL Server uses a syntax similar to SQLite for the `ORDER BY` clause for result sorting within certain aggregate functions but instead of using the `OVER` clause, it uses `WITHIN GROUP`. I modified `AggregateExpression` to make the `OVER` clause dynamic so that it could be changed for TSQL.

SQL Server uses `STRING_AGG(field1, ',') WITHIN GROUP (ORDER BY field2 DESC)` rather than
`GROUP_CONCAT(field1, ';') OVER (ORDER BY field2 DESC)` for SQLite.

https://docs.microsoft.com/en-us/sql/t-sql/functions/string-agg-transact-sql?view=sql-server-ver15

Comment (othercorey) [reply src/Database/Expression/AggregateExpression.php:187]: This clause is specifically for window expressions.

File: tests/TestCase/Database/FunctionsBuilderTest.php [hunk 2]
```diff
@@ -40,7 +40,6 @@ class FunctionsBuilderTest extends TestCase
      */
     public function setUp(): void
     {
-        parent::setUp();
```
Comment (markstory) [inline tests/TestCase/Database/FunctionsBuilderTest.php:43]: Why was this necessary?

Comment (jimbo2150) [reply tests/TestCase/Database/FunctionsBuilderTest.php:43]: Thanks, looks like I accidentally deleted that line I moved the tests out of that file.

**Comment by jimbo2150** [comment]
The errors don't appear to have to do with this PR.

Note that the tests will not cover the MS SQL server code since the test server is using SQL Server 2016 (version 13). The STRING_AGG and other aggregate related functions/expressions were added in SQL Server 2017 (version 14).

File: src/Database/Expression/GroupedExpression.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+     * Overwrites or initializes the orderContainer property with a
+     * FunctionExpression object.
+     *
+     * @param bool $initialize If true, will not use an existing value.
+     * @return $this
+     */
+    protected function resetOrderContainer($initialize = false)
+    {
+        $this->orderContainer = (new FunctionExpression($initialize ? 'OVER ' : $this->getOrderContainer()->getName()))
+            ->setConjunction(' ', false, false);
```
Comment (markstory) [inline src/Database/Expression/GroupedExpression.php:57]: Why does this replace the previous container with a new copy?

File: src/Database/Expression/GroupedExpression.php [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+namespace Cake\Database\Expression;
+
+use Cake\Database\ValueBinder;
+use Closure;
+
+/**
+ * Handles the few special aggregate functions that require special syntax
+ * for grouped ordering.
+ */
+class GroupedExpression extends FunctionExpression
```
Comment (markstory) [inline src/Database/Expression/GroupedExpression.php:26]: GroupConcat might be a better name for this based on how it is used.

File: src/Database/Expression/GroupedExpression.php [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    /**
+     * @var \Cake\Database\Expression\OrderByExpression|null
+     */
+    protected $order;
+
+    /**
+     * @inheritDoc
+     */
+    public function __construct(string $name, array $params = [], array $types = [], string $returnType = 'string')
```
Comment (markstory) [inline src/Database/Expression/GroupedExpression.php:41]: If this expression is only used for GroupConcat does the name need to be a parameter?

File: src/Database/Expression/SelectExpression.php [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+    /**
+     * Method that returns the modifiers as either an associative array where
+     * the modifiers are keys or as an array where the modifiers are array
+     * values.
+     *
+     * @param bool $asValue Boolean determining return format.
+     * @return string[]
+     */
+    public function getModifiers(bool $asValue = true): array
```
Comment (markstory) [inline src/Database/Expression/SelectExpression.php:144]: Do we need this parameter? It makes the return value inconsistent and doesn't appear to have any practical usage that the default value doesn't cover.
